src/systems/hover_interaction.rs

```diff
@@ -0,0 +1,107 @@
+use amethyst::core::cgmath::{MetricSpace, Point3};
+use amethyst::core::GlobalTransform;
+use amethyst::ecs::prelude::{
+    Component, DenseVecStorage, Entities, Entity, Join, Read, ReadStorage, System, Write,
+    WriteStorage,
+};
+use amethyst::renderer::Material;
+use collision::primitive::Primitive3;
+use collision::ContinuousTransformed;
+use systems::MouseRay;
+
+pub struct HoverInteractionSystem;
+
+impl<'a> System<'a> for HoverInteractionSystem {
+    type SystemData = (
+        Entities<'a>,
+        Read<'a, MouseRay>,
+        ReadStorage<'a, GlobalTransform>,
+        WriteStorage<'a, HoverHandler>,
+        WriteStorage<'a, Material>,
+        Write<'a, Option<Hovered>>,
+    );
+
+    fn run(
+        &mut self,
+        (entities, mouse_ray, transforms, mut hover_handlers, mut materials, mut hovered): Self::SystemData,
+    ) {
+        let mut nearest = None;
+        for (entity, hover_handler, transform) in
+            (&*entities, &mut hover_handlers, &transforms).join()
+        {
+            if let Some(point) = hover_handler
+                .bounding_box
+                .intersection_transformed(&mouse_ray.ray, &transform.0)
+            {
+                let collision_distance = mouse_ray.ray.origin.distance2(point);
+                if nearest.map_or(true, |(nearest_distance, _, _)| {
+                    collision_distance < nearest_distance
+                }) {
+                    nearest = Some((collision_distance, point, entity))
+                }
+            }
+        }
+
+        let old_hovered_entity = (*hovered).clone();
+
+        *hovered = nearest.map(|(_, point, entity)| Hovered {
+            entity: entity,
+            point: point,
+        });
+
+        match (&old_hovered_entity, &*hovered) {
+            (Some(Hovered { entity: a, .. }), Some(Hovered { entity: b, .. })) if a != b => {
```

Comment:
```
        old_hovered_entity.map(|hovered|hover_handlers
            .get_mut(hovered.entity)
            .unwrap()
            .change_materials(&hovered.entity, &mut materials));

        // we cannot use `map()` here, because map would move `hovered` while only only borrowed it from the world
        if let Some(Hovered{entity:e, ..}) = *hovered {
            hover_handlers
                .get_mut(e)
                .unwrap()
                .change_materials(&e, &mut materials)
        }```

Reply:
would look like this